### PR TITLE
Full RNN support with control flow ops

### PR DESCRIPTION
**Pull request by roywei**
### Summary

- Add control flow ops in RNN so we can set unroll=False as default
- resolved drop out issue: https://github.com/awslabs/keras-apache-mxnet/issues/60
- enabled 18 rnn unit tests

### Related Issues
https://github.com/awslabs/keras-apache-mxnet/wiki/Using-RNN-with-MXNet-backend
https://github.com/awslabs/keras-apache-mxnet/issues/60

### PR Overview

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [x] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [x] This PR changes the current API [y/n]

### CODE REVIEW

File: keras/backend/mxnet_backend.py [hunk 1]
```diff
@@ -2689,56 +2674,87 @@ def rnn(step_function, inputs, initial_states,
...
+                last_output = KerasSymbol(mx.sym.where(msk,
+                                                       last_output.symbol,
+                                                       prev_output.symbol))
+                prev_output = last_output
+            states = new_states
+            # Expand the output dimension from `(samples, output_dim)` to
+            # `(samples, 1, output_dim)` with middle axis as time dimension
+            outputs.append(mx.sym.expand_dims(last_output.symbol, axis=1))
+        # Concatenate the output across time dimension
+        outputs = mx.sym.Concat(*outputs, dim=1)
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: `mx.sym.Concat` has been deprecated, use `mx.sym.concat` instead

File: keras/backend/mxnet_backend.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 import warnings
 import mxnet as mx
 import numpy as np
+import keras
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: why are we importing the entire library? Please import only the specific modules that you have used in the PR

File: keras/backend/mxnet_backend.py [hunk 3]
```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
...
+            if not states:
+                raise ValueError('MXNet Backend: Initial states is not provided when masking is '
+                                 'enabled.')
+            if mask.dtype != dtype:
+                mask = cast(mask, dtype)
+            # Split the mask across time dimension and generate the list of masks
+            # with shape `(samples, 1)` (no time dimension)
+            masks = list(mx.sym.split(mask.symbol, axis=1,
+                                      squeeze_axis=1, num_outputs=dshape[1]))
+            # Reverse the mask sequence
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: remove comment

File: keras/backend/mxnet_backend.py [hunk 4]
```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
...
+            if go_backwards:
+                masks.reverse()
+        else:
+            masks = [None for _ in inputs]
+
+        # Iterate over a time step
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: remove comment

File: keras/backend/mxnet_backend.py [hunk 5]
```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
...
+                    prev_output = zeros_like(last_output)
+                last_output = KerasSymbol(mx.sym.where(msk,
+                                                       last_output.symbol,
+                                                       prev_output.symbol))
+                prev_output = last_output
+            states = new_states
+            # Expand the output dimension from `(samples, output_dim)` to
+            # `(samples, 1, output_dim)` with middle axis as time dimension
+            outputs.append(mx.sym.expand_dims(last_output.symbol, axis=1))
+        # Concatenate the output across time dimension
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: remove comment

File: keras/backend/mxnet_backend.py [hunk 6]
```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
...
+                dp_mask = None
+                # dropout matrices for recurrent units
+                rec_dp_mask = None
+
+                if 0 < cell.dropout < 1 and cell._dropout_mask is None:
+                    dp_mask = _generate_dropout_mask(
+                        KerasSymbol(mx.sym.ones_like(inputs)),
+                        cell.dropout,
+                        training=training)
+                if (0 < cell.recurrent_dropout < 1 and
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: remove parenthesis

File: keras/backend/mxnet_backend.py [hunk 7]
```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
...
+                        h_tm1_o = h_tm1
+
+                    i = cell.recurrent_activation(KerasSymbol(x_i +
+                                                              mx.sym.dot(h_tm1_i,
+                                                                         cell.recurrent_kernel_i.symbol))).symbol
+                    f = cell.recurrent_activation(KerasSymbol(x_f +
+                                                              mx.sym.dot(h_tm1_f,
+                                                                         cell.recurrent_kernel_f.symbol))).symbol
+                    c = f * c_tm1 + i * cell.activation(KerasSymbol(x_c +
+                                                                    mx.sym.dot(h_tm1_c,
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: [minor] indentation

File: keras/backend/mxnet_backend.py [hunk 8]
```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
...
+                # dropout matrices for recurrent units
+                rec_dp_mask = None
+
+                if 0 < cell.dropout < 1 and cell._dropout_mask is None:
+                    dp_mask = _generate_dropout_mask(
+                        KerasSymbol(mx.sym.ones_like(inputs)),
+                        cell.dropout,
+                        training=training,
+                        count=3)
+                if (0 < cell.recurrent_dropout < 1 and
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: remove paranthesis

File: keras/backend/mxnet_backend.py [hunk 9]
```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
...
+            states = new_states
+            # Expand the output dimension from `(samples, output_dim)` to
+            # `(samples, 1, output_dim)` with middle axis as time dimension
+            outputs.append(mx.sym.expand_dims(last_output.symbol, axis=1))
+        # Concatenate the output across time dimension
+        outputs = mx.sym.concat(*outputs, dim=1)
     else:
-        masks = [None for _ in inputs]
+        # TODO: remove version check after mxnet 1.3.1 release
+        if mx.__version__ != '1.3.1':
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: add unit test to check for this condition

Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: This may not work always because you can have .post0 etc. releases. Please use regex for guaranteed checks

File: keras/backend/mxnet_backend.py [hunk 10]
```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
...
+        else:
+            data = [inputs]
+        states = [state.symbol for state in initial_states] + [constant.symbol for constant in constants]
+
+        # using control flow operators(foreach) in MXNet if not unrolling
+        # foreach operator only take step functions with MXNet symbol as input
+        # translating step function from Keras RNN cells to pure MXNet
+        # define _step according to different type of RNN Cells
+        if type(cell) == keras.layers.recurrent.SimpleRNNCell:
+            def _step(data, states):
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: Please move these to separate functions for easy maintenance.

def _step_for_simple_rnn(), _step_for_lstm() .. etc.

File: keras/backend/mxnet_backend.py [hunk 11]
```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
...
+        outputs = mx.sym.transpose(outputs, axes)
+        # Properly set learning phase on output tensor.
+    outputs = KerasSymbol(outputs)
     last_output._uses_learning_phase = uses_learning_phase
-    return last_output, KerasSymbol(outputs), states
+    return last_output, outputs, states
+
+
+def _dot_rnn(x, y):
+    ndim_y = len(y.infer_shape_partial()[1][0])
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: y.infer_shape_partial()[1] can be None

File: keras/backend/mxnet_backend.py [hunk 12]
```diff
@@ -4920,9 +5223,17 @@ def _sync_weights(self):
             if self._weights_dirty:
                 args, auxs = self._module.get_params()
                 for name in self._arg_names:
-                    self._args[name][:] = args[name]
+                    try:
+                        self._args[name][:] = args[name]
+                    except:
+                        self._args[name] = []
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: why is this required?

File: keras/backend/mxnet_backend.py (hunk 5 above)
Comment (roywei) [reply keras/backend/mxnet_backend.py]: it's important to explain we are concatenate on time dimension (dim=1)

File: keras/backend/mxnet_backend.py (hunk 3 above)
Comment (roywei) [reply keras/backend/mxnet_backend.py]: This is from original rnn implementation with unrolling, let's keep it

File: keras/backend/mxnet_backend.py (hunk 4 above)
Comment (roywei) [reply keras/backend/mxnet_backend.py]: This is from original rnn implementation with unrolling, let's keep it

**Comment by kalyc** [comment]
Will this require update in documentation? `using_rnn_with_mxnet_backend`?

**Comment by roywei** [comment]
> Will this require update in documentation? `using_rnn_with_mxnet_backend`?

Updated

File: docs/mxnet_backend/using_rnn_with_mxnet_backend.md [hunk 13]
```diff
@@ -57,9 +53,13 @@ NOTE:
     arbitrary large maxlen. It is always optimal to choose maxlen for padding to be equal to the max length of the 
     input sequences.
 ```
-## Unroll=False is not supported
+## Using unrolling in RNN
+
+In Keras RNN layers, by default unroll is set to False, and it requires control flow opreators(e.g. foreach, while_loop).
```
Comment (kalyc) [inline docs/mxnet_backend/using_rnn_with_mxnet_backend.md]: operators ** spelling

File: docs/mxnet_backend/using_rnn_with_mxnet_backend.md [hunk 14]
```diff
@@ -57,9 +53,13 @@ NOTE:
...
+We recently added support for this feature so by default RNN layers are not unrolled, same as other backends.
 
-As described above, MXNet backend does not support variable length input and hence `unroll=False` is not supported in RNN layers. You are expected to provide `input_shape` and set `unroll=True`.
+Unrolling RNN Cells will have better performance but comsumes more memory. It's only suitable for short sequeence. For more details, refer to
```
Comment (kalyc) [inline docs/mxnet_backend/using_rnn_with_mxnet_backend.md]: sequence ** spelling
